{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4f5ef54b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqIO as si"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6aab0d4",
   "metadata": {},
   "source": [
    "### This script creates a list of 1:1:1 orthologs gene between human, mouse and bat (Rousettus aegyptiacus).\n",
    "### The script takes CDS fasta file and EggNOG's run results for each species and returns a dataframe containing all genes that have an identical EggNOG gene name."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f51b8a36",
   "metadata": {},
   "source": [
    "#### Loading and Extracting genes data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "97a5cf50",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load mouse annotation file from Eggnog run results.\n",
    "mouse_annotations = pd.read_csv(r\"C:\\Users\\TzachiHNB6\\Documents\\eggnog\\mouse_query_92_1.fa.emapper.annotations\", delimiter=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7b4e689b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load human annotation file from Eggnog run results.\n",
    "human_annotations  = pd.read_csv(r\"C:\\Users\\TzachiHNB6\\Documents\\eggnog\\human_query_92_1.fa.emapper.annotations\", delimiter=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9b84cdaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load bat annotation file from Eggnog run results.\n",
    "bat1k_annotations = pd.read_csv(r\"C:\\Users\\TzachiHNB6\\Documents\\eggnog\\bat1k_query_1.fa.emapper.annotations\", delimiter=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "470e1efd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This function goes over all cds record and extracts record id (that will be used for intersection with Eggnog annotations file)\n",
    "# gene id and gene name (if exists)\n",
    "def extract_cds_data(cds_path, gene_id_regex = None, gene_name_regex = None):\n",
    "    gene_ids = []\n",
    "    gene_names = []\n",
    "    fasta_record_ids = []\n",
    "    with open(cds_path, \"r\") as cds:\n",
    "        sequences = si.parse(cds, \"fasta\")\n",
    "        for fasta in sequences:\n",
    "            fasta_record_ids.append(fasta.id)\n",
    "            if gene_id_regex != None:\n",
    "                gene_ids.append(re.search(gene_id_regex, fasta.description).group(1))\n",
    "            if gene_name_regex != None:\n",
    "                gene_names.append(re.search(gene_name_regex, fasta.description).group(1))\n",
    "    if gene_id_regex != None and gene_name_regex != None:\n",
    "        return pd.DataFrame({'record_id' : fasta_record_ids, 'gene_id': gene_ids, 'gene_name': gene_names})\n",
    "    if gene_id_regex != None:\n",
    "        return pd.DataFrame({'record_id' : fasta_record_ids, 'gene_id': gene_ids})\n",
    "    return pd.DataFrame({'record_id' : fasta_record_ids, 'gene_name': gene_names})\n",
    "    \n",
    "\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "875cc99b",
   "metadata": {},
   "outputs": [],
   "source": [
    "cds_path = r\"C:\\Users\\TzachiHNB6\\Documents\\eggnog\\genoms\\mouse_92_cds\\Mus_musculus.GRCm38.cds.all.fa\"\n",
    "mouse_data = extract_cds_data(cds_path, \"gene:([^ ]+)\", \"gene_symbol:([^ ]+)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "b196a5ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>record_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUST00000196221.1</td>\n",
       "      <td>ENSMUSG00000096749.2</td>\n",
       "      <td>Trdd1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUST00000177564.1</td>\n",
       "      <td>ENSMUSG00000096176.1</td>\n",
       "      <td>Trdd2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUST00000178537.1</td>\n",
       "      <td>ENSMUSG00000095668.1</td>\n",
       "      <td>Trbd1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUST00000178862.1</td>\n",
       "      <td>ENSMUSG00000094569.1</td>\n",
       "      <td>Trbd2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUST00000179520.1</td>\n",
       "      <td>ENSMUSG00000094028.1</td>\n",
       "      <td>Ighd4-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65465</th>\n",
       "      <td>ENSMUST00000201577.1</td>\n",
       "      <td>ENSMUSG00000107104.3</td>\n",
       "      <td>Nrxn2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65466</th>\n",
       "      <td>ENSMUST00000201188.3</td>\n",
       "      <td>ENSMUSG00000107104.3</td>\n",
       "      <td>Nrxn2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65467</th>\n",
       "      <td>ENSMUST00000200719.1</td>\n",
       "      <td>ENSMUSG00000107104.3</td>\n",
       "      <td>Nrxn2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65468</th>\n",
       "      <td>ENSMUST00000202867.3</td>\n",
       "      <td>ENSMUSG00000107099.3</td>\n",
       "      <td>Slc22a12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65469</th>\n",
       "      <td>ENSMUST00000202211.1</td>\n",
       "      <td>ENSMUSG00000107099.3</td>\n",
       "      <td>Slc22a12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>65470 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  record_id               gene_id gene_name\n",
       "0      ENSMUST00000196221.1  ENSMUSG00000096749.2     Trdd1\n",
       "1      ENSMUST00000177564.1  ENSMUSG00000096176.1     Trdd2\n",
       "2      ENSMUST00000178537.1  ENSMUSG00000095668.1     Trbd1\n",
       "3      ENSMUST00000178862.1  ENSMUSG00000094569.1     Trbd2\n",
       "4      ENSMUST00000179520.1  ENSMUSG00000094028.1   Ighd4-1\n",
       "...                     ...                   ...       ...\n",
       "65465  ENSMUST00000201577.1  ENSMUSG00000107104.3     Nrxn2\n",
       "65466  ENSMUST00000201188.3  ENSMUSG00000107104.3     Nrxn2\n",
       "65467  ENSMUST00000200719.1  ENSMUSG00000107104.3     Nrxn2\n",
       "65468  ENSMUST00000202867.3  ENSMUSG00000107099.3  Slc22a12\n",
       "65469  ENSMUST00000202211.1  ENSMUSG00000107099.3  Slc22a12\n",
       "\n",
       "[65470 rows x 3 columns]"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mouse_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "12cf360e",
   "metadata": {},
   "outputs": [],
   "source": [
    "cds_path = r\"C:\\Users\\TzachiHNB6\\Documents\\eggnog\\genoms\\human_92_cds\\Homo_sapiens.GRCh38.cds.all.fa\"\n",
    "human_data = extract_cds_data(cds_path, \"gene:([^ ]+)\", \"gene_symbol:([^ ]+)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "0bd91b25",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>record_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000434970.2</td>\n",
       "      <td>ENSG00000237235.2</td>\n",
       "      <td>TRDD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000448914.1</td>\n",
       "      <td>ENSG00000228985.1</td>\n",
       "      <td>TRDD3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000415118.1</td>\n",
       "      <td>ENSG00000223997.1</td>\n",
       "      <td>TRDD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000632684.1</td>\n",
       "      <td>ENSG00000282431.1</td>\n",
       "      <td>TRBD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000631435.1</td>\n",
       "      <td>ENSG00000282253.1</td>\n",
       "      <td>TRBD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107839</th>\n",
       "      <td>ENST00000645885.1</td>\n",
       "      <td>ENSG00000285075.1</td>\n",
       "      <td>TPK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107840</th>\n",
       "      <td>ENST00000642257.1</td>\n",
       "      <td>ENSG00000285075.1</td>\n",
       "      <td>TPK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107841</th>\n",
       "      <td>ENST00000643965.1</td>\n",
       "      <td>ENSG00000285075.1</td>\n",
       "      <td>TPK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107842</th>\n",
       "      <td>ENST00000642712.1</td>\n",
       "      <td>ENSG00000285114.1</td>\n",
       "      <td>GSDMC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107843</th>\n",
       "      <td>ENST00000646090.1</td>\n",
       "      <td>ENSG00000285114.1</td>\n",
       "      <td>GSDMC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>107844 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                record_id            gene_id gene_name\n",
       "0       ENST00000434970.2  ENSG00000237235.2     TRDD2\n",
       "1       ENST00000448914.1  ENSG00000228985.1     TRDD3\n",
       "2       ENST00000415118.1  ENSG00000223997.1     TRDD1\n",
       "3       ENST00000632684.1  ENSG00000282431.1     TRBD1\n",
       "4       ENST00000631435.1  ENSG00000282253.1     TRBD1\n",
       "...                   ...                ...       ...\n",
       "107839  ENST00000645885.1  ENSG00000285075.1      TPK1\n",
       "107840  ENST00000642257.1  ENSG00000285075.1      TPK1\n",
       "107841  ENST00000643965.1  ENSG00000285075.1      TPK1\n",
       "107842  ENST00000642712.1  ENSG00000285114.1     GSDMC\n",
       "107843  ENST00000646090.1  ENSG00000285114.1     GSDMC\n",
       "\n",
       "[107844 rows x 3 columns]"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "2eee915d",
   "metadata": {},
   "outputs": [],
   "source": [
    "cds_path = r\"C:\\Users\\TzachiHNB6\\Documents\\eggnog\\genoms\\bat_cds\\cds_from_genomic.fna\"\n",
    "bat_data = extract_cds_data(cds_path, gene_name_regex = \"gene=([^\\]]+)]\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "fa9d38fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>record_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lcl|NW_023416284.1_cds_XP_016000028.1_1</td>\n",
       "      <td>SAMD11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lcl|NW_023416284.1_cds_XP_016000029.1_2</td>\n",
       "      <td>SAMD11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lcl|NW_023416284.1_cds_XP_016000030.1_3</td>\n",
       "      <td>SAMD11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lcl|NW_023416284.1_cds_XP_016000033.2_4</td>\n",
       "      <td>NOC2L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lcl|NW_023416284.1_cds_XP_016000032.2_5</td>\n",
       "      <td>NOC2L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56471</th>\n",
       "      <td>lcl|NC_007393.1_cds_YP_313613.1_49055</td>\n",
       "      <td>ND4L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56472</th>\n",
       "      <td>lcl|NC_007393.1_cds_YP_313614.1_49056</td>\n",
       "      <td>ND4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56473</th>\n",
       "      <td>lcl|NC_007393.1_cds_YP_313615.1_49057</td>\n",
       "      <td>ND5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56474</th>\n",
       "      <td>lcl|NC_007393.1_cds_YP_313616.1_49058</td>\n",
       "      <td>ND6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56475</th>\n",
       "      <td>lcl|NC_007393.1_cds_YP_313617.1_49059</td>\n",
       "      <td>CYTB</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>56476 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     record_id gene_name\n",
       "0      lcl|NW_023416284.1_cds_XP_016000028.1_1    SAMD11\n",
       "1      lcl|NW_023416284.1_cds_XP_016000029.1_2    SAMD11\n",
       "2      lcl|NW_023416284.1_cds_XP_016000030.1_3    SAMD11\n",
       "3      lcl|NW_023416284.1_cds_XP_016000033.2_4     NOC2L\n",
       "4      lcl|NW_023416284.1_cds_XP_016000032.2_5     NOC2L\n",
       "...                                        ...       ...\n",
       "56471    lcl|NC_007393.1_cds_YP_313613.1_49055      ND4L\n",
       "56472    lcl|NC_007393.1_cds_YP_313614.1_49056       ND4\n",
       "56473    lcl|NC_007393.1_cds_YP_313615.1_49057       ND5\n",
       "56474    lcl|NC_007393.1_cds_YP_313616.1_49058       ND6\n",
       "56475    lcl|NC_007393.1_cds_YP_313617.1_49059      CYTB\n",
       "\n",
       "[56476 rows x 2 columns]"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bat_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5dd14d0",
   "metadata": {},
   "source": [
    "#### Combine with eggnog results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "340d7202",
   "metadata": {},
   "outputs": [],
   "source": [
    "def combine_with_eggnog(data,annotations, gene_id_col_name = \"record_id\"):\n",
    "    for index,row in data.iterrows():\n",
    "        values = annotations.loc[annotations[\"query\"] == row[gene_id_col_name], \"Preferred_name\"].values\n",
    "        if len(values) > 0:\n",
    "            data.loc[index,\"eggnog_name\"] =  values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "97cafa8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "combine_with_eggnog(bat_data, bat1k_annotations)\n",
    "combine_with_eggnog(human_data, human_annotations)\n",
    "combine_with_eggnog(mouse_data, mouse_annotations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "378b452c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>record_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>eggnog_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lcl|NW_023416284.1_cds_XP_016000028.1_1</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>SAMD11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lcl|NW_023416284.1_cds_XP_016000029.1_2</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>SAMD11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lcl|NW_023416284.1_cds_XP_016000030.1_3</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>SAMD11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lcl|NW_023416284.1_cds_XP_016000033.2_4</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NOC2L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lcl|NW_023416284.1_cds_XP_016000032.2_5</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NOC2L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56471</th>\n",
       "      <td>lcl|NC_007393.1_cds_YP_313613.1_49055</td>\n",
       "      <td>ND4L</td>\n",
       "      <td>ND4L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56472</th>\n",
       "      <td>lcl|NC_007393.1_cds_YP_313614.1_49056</td>\n",
       "      <td>ND4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56473</th>\n",
       "      <td>lcl|NC_007393.1_cds_YP_313615.1_49057</td>\n",
       "      <td>ND5</td>\n",
       "      <td>ND5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56474</th>\n",
       "      <td>lcl|NC_007393.1_cds_YP_313616.1_49058</td>\n",
       "      <td>ND6</td>\n",
       "      <td>ND6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56475</th>\n",
       "      <td>lcl|NC_007393.1_cds_YP_313617.1_49059</td>\n",
       "      <td>CYTB</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>56476 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     record_id gene_name eggnog_name\n",
       "0      lcl|NW_023416284.1_cds_XP_016000028.1_1    SAMD11      SAMD11\n",
       "1      lcl|NW_023416284.1_cds_XP_016000029.1_2    SAMD11      SAMD11\n",
       "2      lcl|NW_023416284.1_cds_XP_016000030.1_3    SAMD11      SAMD11\n",
       "3      lcl|NW_023416284.1_cds_XP_016000033.2_4     NOC2L       NOC2L\n",
       "4      lcl|NW_023416284.1_cds_XP_016000032.2_5     NOC2L       NOC2L\n",
       "...                                        ...       ...         ...\n",
       "56471    lcl|NC_007393.1_cds_YP_313613.1_49055      ND4L        ND4L\n",
       "56472    lcl|NC_007393.1_cds_YP_313614.1_49056       ND4         NaN\n",
       "56473    lcl|NC_007393.1_cds_YP_313615.1_49057       ND5         ND5\n",
       "56474    lcl|NC_007393.1_cds_YP_313616.1_49058       ND6         ND6\n",
       "56475    lcl|NC_007393.1_cds_YP_313617.1_49059      CYTB         NaN\n",
       "\n",
       "[56476 rows x 3 columns]"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bat_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "id": "cd971292",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>record_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>eggnog_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000434970.2</td>\n",
       "      <td>ENSG00000237235.2</td>\n",
       "      <td>TRDD2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000448914.1</td>\n",
       "      <td>ENSG00000228985.1</td>\n",
       "      <td>TRDD3</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000415118.1</td>\n",
       "      <td>ENSG00000223997.1</td>\n",
       "      <td>TRDD1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000632684.1</td>\n",
       "      <td>ENSG00000282431.1</td>\n",
       "      <td>TRBD1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000631435.1</td>\n",
       "      <td>ENSG00000282253.1</td>\n",
       "      <td>TRBD1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107839</th>\n",
       "      <td>ENST00000645885.1</td>\n",
       "      <td>ENSG00000285075.1</td>\n",
       "      <td>TPK1</td>\n",
       "      <td>TPK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107840</th>\n",
       "      <td>ENST00000642257.1</td>\n",
       "      <td>ENSG00000285075.1</td>\n",
       "      <td>TPK1</td>\n",
       "      <td>TPK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107841</th>\n",
       "      <td>ENST00000643965.1</td>\n",
       "      <td>ENSG00000285075.1</td>\n",
       "      <td>TPK1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107842</th>\n",
       "      <td>ENST00000642712.1</td>\n",
       "      <td>ENSG00000285114.1</td>\n",
       "      <td>GSDMC</td>\n",
       "      <td>GSDMC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107843</th>\n",
       "      <td>ENST00000646090.1</td>\n",
       "      <td>ENSG00000285114.1</td>\n",
       "      <td>GSDMC</td>\n",
       "      <td>GSDMC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>107844 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                record_id            gene_id gene_name eggnog_name\n",
       "0       ENST00000434970.2  ENSG00000237235.2     TRDD2         NaN\n",
       "1       ENST00000448914.1  ENSG00000228985.1     TRDD3         NaN\n",
       "2       ENST00000415118.1  ENSG00000223997.1     TRDD1         NaN\n",
       "3       ENST00000632684.1  ENSG00000282431.1     TRBD1         NaN\n",
       "4       ENST00000631435.1  ENSG00000282253.1     TRBD1         NaN\n",
       "...                   ...                ...       ...         ...\n",
       "107839  ENST00000645885.1  ENSG00000285075.1      TPK1        TPK1\n",
       "107840  ENST00000642257.1  ENSG00000285075.1      TPK1        TPK1\n",
       "107841  ENST00000643965.1  ENSG00000285075.1      TPK1         NaN\n",
       "107842  ENST00000642712.1  ENSG00000285114.1     GSDMC       GSDMC\n",
       "107843  ENST00000646090.1  ENSG00000285114.1     GSDMC       GSDMC\n",
       "\n",
       "[107844 rows x 4 columns]"
      ]
     },
     "execution_count": 211,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "id": "900e5a75",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>record_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>eggnog_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUST00000196221.1</td>\n",
       "      <td>ENSMUSG00000096749.2</td>\n",
       "      <td>Trdd1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUST00000177564.1</td>\n",
       "      <td>ENSMUSG00000096176.1</td>\n",
       "      <td>Trdd2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUST00000178537.1</td>\n",
       "      <td>ENSMUSG00000095668.1</td>\n",
       "      <td>Trbd1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUST00000178862.1</td>\n",
       "      <td>ENSMUSG00000094569.1</td>\n",
       "      <td>Trbd2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUST00000179520.1</td>\n",
       "      <td>ENSMUSG00000094028.1</td>\n",
       "      <td>Ighd4-1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65465</th>\n",
       "      <td>ENSMUST00000201577.1</td>\n",
       "      <td>ENSMUSG00000107104.3</td>\n",
       "      <td>Nrxn2</td>\n",
       "      <td>NRXN2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65466</th>\n",
       "      <td>ENSMUST00000201188.3</td>\n",
       "      <td>ENSMUSG00000107104.3</td>\n",
       "      <td>Nrxn2</td>\n",
       "      <td>NRXN2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65467</th>\n",
       "      <td>ENSMUST00000200719.1</td>\n",
       "      <td>ENSMUSG00000107104.3</td>\n",
       "      <td>Nrxn2</td>\n",
       "      <td>NRXN2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65468</th>\n",
       "      <td>ENSMUST00000202867.3</td>\n",
       "      <td>ENSMUSG00000107099.3</td>\n",
       "      <td>Slc22a12</td>\n",
       "      <td>SLC22A12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65469</th>\n",
       "      <td>ENSMUST00000202211.1</td>\n",
       "      <td>ENSMUSG00000107099.3</td>\n",
       "      <td>Slc22a12</td>\n",
       "      <td>SLC22A12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>65470 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  record_id               gene_id gene_name eggnog_name\n",
       "0      ENSMUST00000196221.1  ENSMUSG00000096749.2     Trdd1         NaN\n",
       "1      ENSMUST00000177564.1  ENSMUSG00000096176.1     Trdd2         NaN\n",
       "2      ENSMUST00000178537.1  ENSMUSG00000095668.1     Trbd1         NaN\n",
       "3      ENSMUST00000178862.1  ENSMUSG00000094569.1     Trbd2         NaN\n",
       "4      ENSMUST00000179520.1  ENSMUSG00000094028.1   Ighd4-1         NaN\n",
       "...                     ...                   ...       ...         ...\n",
       "65465  ENSMUST00000201577.1  ENSMUSG00000107104.3     Nrxn2       NRXN2\n",
       "65466  ENSMUST00000201188.3  ENSMUSG00000107104.3     Nrxn2       NRXN2\n",
       "65467  ENSMUST00000200719.1  ENSMUSG00000107104.3     Nrxn2       NRXN2\n",
       "65468  ENSMUST00000202867.3  ENSMUSG00000107099.3  Slc22a12    SLC22A12\n",
       "65469  ENSMUST00000202211.1  ENSMUSG00000107099.3  Slc22a12    SLC22A12\n",
       "\n",
       "[65470 rows x 4 columns]"
      ]
     },
     "execution_count": 212,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mouse_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5d07398",
   "metadata": {},
   "source": [
    "#### Clean data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "bcba8d81",
   "metadata": {},
   "outputs": [],
   "source": [
    "bat_data_copy = bat_data.copy()\n",
    "human_data_copy = human_data.copy()\n",
    "mouse_data_copy = mouse_data.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "id": "39609756",
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_data(data):\n",
    "    # run upper-case on gene names and remove redundent column\n",
    "    data[\"eggnog_name\"] = [x.upper() for x in data[\"eggnog_name\"]]\n",
    "    data[\"gene_name\"] = [x.upper() for x in data[\"gene_name\"]]\n",
    "    data.drop(columns = [\"record_id\"], inplace=True)\n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "id": "15fc6979",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\TZACHI~1\\AppData\\Local\\Temp/ipykernel_3292/1440718062.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  data[\"eggnog_name\"] = [x.upper() for x in data[\"eggnog_name\"]]\n",
      "C:\\Users\\TZACHI~1\\AppData\\Local\\Temp/ipykernel_3292/1440718062.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  data[\"gene_name\"] = [x.upper() for x in data[\"gene_name\"]]\n",
      "C:\\Users\\TzachiHNB6\\AppData\\Local\\Programs\\Python\\Python39\\lib\\site-packages\\pandas\\core\\frame.py:4906: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  return super().drop(\n"
     ]
    }
   ],
   "source": [
    "bat_data_copy = clean_data(bat_data_copy)\n",
    "mouse_data_copy = clean_data(mouse_data_copy)\n",
    "human_data_copy = clean_data(human_data_copy)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "id": "1fc1e423",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>eggnog_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMD11</td>\n",
       "      <td>SAMD11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMD11</td>\n",
       "      <td>SAMD11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMD11</td>\n",
       "      <td>SAMD11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NOC2L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NOC2L</td>\n",
       "      <td>NOC2L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56466</th>\n",
       "      <td>COX2</td>\n",
       "      <td>COX2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56468</th>\n",
       "      <td>ATP6</td>\n",
       "      <td>ATP6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56471</th>\n",
       "      <td>ND4L</td>\n",
       "      <td>ND4L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56473</th>\n",
       "      <td>ND5</td>\n",
       "      <td>ND5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56474</th>\n",
       "      <td>ND6</td>\n",
       "      <td>ND6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54210 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_name eggnog_name\n",
       "0        SAMD11      SAMD11\n",
       "1        SAMD11      SAMD11\n",
       "2        SAMD11      SAMD11\n",
       "3         NOC2L       NOC2L\n",
       "4         NOC2L       NOC2L\n",
       "...         ...         ...\n",
       "56466      COX2        COX2\n",
       "56468      ATP6        ATP6\n",
       "56471      ND4L        ND4L\n",
       "56473       ND5         ND5\n",
       "56474       ND6         ND6\n",
       "\n",
       "[54210 rows x 2 columns]"
      ]
     },
     "execution_count": 191,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bat_data_copy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "id": "f5c5df04",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove version from gene id in human and mouse\n",
    "mouse_data_copy[\"gene_id\"] = [x.split(\".\")[0] for x in mouse_data_copy[\"gene_id\"]]\n",
    "human_data_copy[\"gene_id\"] = [x.split(\".\")[0] for x in human_data_copy[\"gene_id\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "id": "7841af3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove any gene that has multiple gene ids or has no eggnog name\n",
    "def remove_multiple_gene_ids(data, gene_id_col = \"gene_id\"):\n",
    "    data = data[(data[\"eggnog_name\"].notnull())& (data[\"eggnog_name\"] != \"-\")]\n",
    "    data_grouped = data.groupby(\"eggnog_name\").agg(set).reset_index() # group by eggnog name\n",
    "    data_grouped = data_grouped[data_grouped[gene_id_col].map(len) == 1] # remove genes with more than 1 eggnog name\n",
    "    data_grouped[gene_id_col] = [list(x)[0] for x in data_grouped[gene_id_col]]\n",
    "    return data_grouped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "id": "790d8d74",
   "metadata": {},
   "outputs": [],
   "source": [
    "#remove_multiple_gene_ids(bat_data_copy,\"gene_name\")\n",
    "bat_data_copy = remove_multiple_gene_ids(bat_data_copy,\"gene_name\")\n",
    "human_data_copy = remove_multiple_gene_ids(human_data_copy,\"gene_id\")\n",
    "mouse_data_copy  = remove_multiple_gene_ids(mouse_data_copy,\"gene_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "id": "0fcbeef9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# change the gene name from a single string set to a single string\n",
    "human_data_copy[\"gene_name\"] = [list(x)[0] for x in human_data_copy[\"gene_name\"]]\n",
    "mouse_data_copy[\"gene_name\"] = [list(x)[0] for x in mouse_data_copy[\"gene_name\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "id": "501cbd51",
   "metadata": {},
   "outputs": [],
   "source": [
    "#change gene_name and gene_id column names to be species specific\n",
    "bat_data_copy.rename(columns={\"gene_name\": \"bat_gene_name\"}, inplace=True)\n",
    "human_data_copy.rename(columns={\"gene_name\": \"human_gene_name\", \"gene_id\":\"human_gene_id\"}, inplace=True)\n",
    "mouse_data_copy.rename(columns={\"gene_name\": \"mouse_gene_name\", \"gene_id\":\"mouse_gene_id\"}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "id": "87b1b4c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eggnog_name</th>\n",
       "      <th>mouse_gene_id</th>\n",
       "      <th>mouse_gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>ENSMUSG00000022347</td>\n",
       "      <td>A1BG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENSMUSG00000052595</td>\n",
       "      <td>A1CF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSMUSG00000030111</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>ENSMUSG00000028794</td>\n",
       "      <td>A3GALT2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>ENSMUSG00000047878</td>\n",
       "      <td>A4GALT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17466</th>\n",
       "      <td>ZYG11A</td>\n",
       "      <td>ENSMUSG00000034645</td>\n",
       "      <td>ZYG11A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17467</th>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>ENSMUSG00000034636</td>\n",
       "      <td>ZYG11B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17468</th>\n",
       "      <td>ZYX</td>\n",
       "      <td>ENSMUSG00000029860</td>\n",
       "      <td>ZYX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17469</th>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>ENSMUSG00000055670</td>\n",
       "      <td>ZZEF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17470</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>ENSMUSG00000039068</td>\n",
       "      <td>ZZZ3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16344 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      eggnog_name       mouse_gene_id mouse_gene_name\n",
       "0            A1BG  ENSMUSG00000022347            A1BG\n",
       "1            A1CF  ENSMUSG00000052595            A1CF\n",
       "2             A2M  ENSMUSG00000030111             A2M\n",
       "3         A3GALT2  ENSMUSG00000028794         A3GALT2\n",
       "4          A4GALT  ENSMUSG00000047878          A4GALT\n",
       "...           ...                 ...             ...\n",
       "17466      ZYG11A  ENSMUSG00000034645          ZYG11A\n",
       "17467      ZYG11B  ENSMUSG00000034636          ZYG11B\n",
       "17468         ZYX  ENSMUSG00000029860             ZYX\n",
       "17469       ZZEF1  ENSMUSG00000055670           ZZEF1\n",
       "17470        ZZZ3  ENSMUSG00000039068            ZZZ3\n",
       "\n",
       "[16344 rows x 3 columns]"
      ]
     },
     "execution_count": 208,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mouse_data_copy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e86b0cc5",
   "metadata": {},
   "source": [
    "#### Creating 1:1:1 results file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "id": "ac5d644d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge 3 genes datasets mkeeping only genes that contain an eggnog_name in all species, to get only genes that are 1:1:1 orthologs according to EggNOG DB\n",
    "merged = pd.merge(pd.merge(bat_data_copy,mouse_data_copy,on='eggnog_name', how=\"inner\"),human_data_copy,on='eggnog_name', how=\"inner\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "id": "d7019180",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eggnog_name</th>\n",
       "      <th>bat_gene_name</th>\n",
       "      <th>mouse_gene_id</th>\n",
       "      <th>mouse_gene_name</th>\n",
       "      <th>human_gene_id</th>\n",
       "      <th>human_gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>ENSMUSG00000022347</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>A1BG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENSMUSG00000052595</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>A1CF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A2M</td>\n",
       "      <td>LOC107501662</td>\n",
       "      <td>ENSMUSG00000030111</td>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>ENSMUSG00000028794</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>ENSG00000184389</td>\n",
       "      <td>A3GALT2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>ENSMUSG00000047878</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>ENSG00000128274</td>\n",
       "      <td>A4GALT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13947</th>\n",
       "      <td>ZXDC</td>\n",
       "      <td>ZXDC</td>\n",
       "      <td>ENSMUSG00000034430</td>\n",
       "      <td>ZXDC</td>\n",
       "      <td>ENSG00000070476</td>\n",
       "      <td>ZXDC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13948</th>\n",
       "      <td>ZYG11A</td>\n",
       "      <td>ZYG11A</td>\n",
       "      <td>ENSMUSG00000034645</td>\n",
       "      <td>ZYG11A</td>\n",
       "      <td>ENSG00000203995</td>\n",
       "      <td>ZYG11A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13949</th>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>ENSMUSG00000034636</td>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>ENSG00000162378</td>\n",
       "      <td>ZYG11B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13950</th>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>ENSMUSG00000055670</td>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>ENSG00000074755</td>\n",
       "      <td>ZZEF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13951</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>ENSMUSG00000039068</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>ENSG00000036549</td>\n",
       "      <td>AC118549.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13952 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      eggnog_name bat_gene_name       mouse_gene_id mouse_gene_name  \\\n",
       "0            A1BG          A1BG  ENSMUSG00000022347            A1BG   \n",
       "1            A1CF          A1CF  ENSMUSG00000052595            A1CF   \n",
       "2             A2M  LOC107501662  ENSMUSG00000030111             A2M   \n",
       "3         A3GALT2       A3GALT2  ENSMUSG00000028794         A3GALT2   \n",
       "4          A4GALT        A4GALT  ENSMUSG00000047878          A4GALT   \n",
       "...           ...           ...                 ...             ...   \n",
       "13947        ZXDC          ZXDC  ENSMUSG00000034430            ZXDC   \n",
       "13948      ZYG11A        ZYG11A  ENSMUSG00000034645          ZYG11A   \n",
       "13949      ZYG11B        ZYG11B  ENSMUSG00000034636          ZYG11B   \n",
       "13950       ZZEF1         ZZEF1  ENSMUSG00000055670           ZZEF1   \n",
       "13951        ZZZ3          ZZZ3  ENSMUSG00000039068            ZZZ3   \n",
       "\n",
       "         human_gene_id human_gene_name  \n",
       "0      ENSG00000121410            A1BG  \n",
       "1      ENSG00000148584            A1CF  \n",
       "2      ENSG00000175899             A2M  \n",
       "3      ENSG00000184389         A3GALT2  \n",
       "4      ENSG00000128274          A4GALT  \n",
       "...                ...             ...  \n",
       "13947  ENSG00000070476            ZXDC  \n",
       "13948  ENSG00000203995          ZYG11A  \n",
       "13949  ENSG00000162378          ZYG11B  \n",
       "13950  ENSG00000074755           ZZEF1  \n",
       "13951  ENSG00000036549      AC118549.1  \n",
       "\n",
       "[13952 rows x 6 columns]"
      ]
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "id": "2e06ac1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save 1:1:1 orthologs to file\n",
    "merged.to_csv(r\"bat_mouse_human_integration\\bat_mouse_human_orthologs.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
